### PR TITLE
O3-4043: Dispensing: Allow back dating dispense events for retrospect…

### PR DESCRIPTION
**Pull request by mogoodrich**
…ive entry

## Requirements

- [ ] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [ ]  My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary

See: https://openmrs.atlassian.net/browse/O3-4043

## Screenshots

image

## Other

Note that this does not yet cover limiting to providers with certain roles, I need to update the REST Web Services module first to support that.

### CODE REVIEW

File: src/history/history-and-comments.component.tsx
Comment (mogoodrich) [inline src/history/history-and-comments.component.tsx:216]: The FHIR resource for a Medication Dispense supports three dates:
* date recorded (maps to date created in the OpenMRS model)
* when prepared (maps to date prepared in the OpenMRS model)
* when handed over (maps to date handed over in the OpenMRS model)

Previously we were timestamping "when prepared" and "when handed over" to "now" (and date_created was timestamped to "now" on the backend).

Now we allow the user to set the "when prepared" and "when handed over" dates via a single Date of Dispense datepicker (which defaults to now).

I don't remember why we were using the date recorded as the display and sort date, but now that we are allowed orders to be back dated, we need to display and sort by when prepared or when handled over, because these are the more accurate and relevant dates.

We are continuing to not distinguish between "date prepared" and "date handed over" and always set them to the  single "dispense date", but the data module will allow us to expand the Dispensing app to record these two potentially distinct events if we like.

File: src/forms/medication-dispense-review.component.tsx
Comment (mseaton) [inline src/forms/medication-dispense-review.component.tsx:502]: What happens here if prescriptionDate = null?  That seems like it is possible from the earlier code

**Review by mseaton — COMMENTED** [review]
One comment, defer to the other react gurus on the rest

File: src/forms/medication-dispense-review.component.tsx
Comment (mogoodrich) [reply src/forms/medication-dispense-review.component.tsx:502]: Oh, interesting, I hadn't really thought of this, but since dayjs() returns the current datetime, I would think dayjs(null) would return the current datetime as well.

Prescription Date should never be null long-term, but as it is loaded via a hook, it might start null.  In this case, having the min date be today until the prescription is properly loaded like proper behavior?

But I also just might want to *not* display the form until all the data is loaded, I can look into that, as it might fix a weird issue I was seeing as well.

File: src/forms/medication-dispense-review.component.tsx
Comment (mogoodrich) [reply src/forms/medication-dispense-review.component.tsx:499]: Thanks @pirupius  I can look into doing that.

File: src/forms/medication-dispense-review.component.tsx
Comment (chibongho) [reply src/forms/medication-dispense-review.component.tsx:502]: > I would think dayjs(null) would return the current datetime as well.

I just checked. It actually creates an "Invalid Date" object if the input is null. It should be fine as long as we have a null check prior

File: src/forms/medication-dispense-review.component.tsx
Comment (mogoodrich) [inline src/forms/medication-dispense-review.component.tsx:504]: @denniskigen  @ibacher  @pirupius  would be great to get your thoughts on whether I'm using this date component correctly, specifically when setting and updating the date.

One weird thing I'm seeing is this display in the datepicker (12:00AM), which I'm guessing happens because I'm passing a full date (with time) in?

image

And also, do we have control of the date display format, or is this controller by the locale?  We try to avoid the American Month-Day-Year, but maybe it is just displaying that because I'm in the US?

image

**Comment by mogoodrich** [comment]
Make some changes to address code review comment, and also now have support for limiting by dispensers by provider role (pending merge of https://openmrs.atlassian.net/browse/PROV-108)

Ready for re-review

File: src/config-schema.ts
Comment (mogoodrich) [reply src/config-schema.ts:59]: Fair enough, even if we go with the comma-separated list on the REST web services side to match other conventions, we should use an array here.

Comment (mogoodrich) [reply src/config-schema.ts]: Fixed!

File: src/forms/medication-dispense-review.component.tsx
Comment (mseaton) [reply src/forms/medication-dispense-review.component.tsx:504]: Let's start a separate ticket / conversation about this @mogoodrich

**Review by chibongho — COMMENTED** [review]
We should resolve the OpnemrsDatePicker, but looks good otherwise.

File: src/forms/medication-dispense-review.component.tsx
Comment (pirupius) [reply src/forms/medication-dispense-review.component.tsx:504]: Yes @mogoodrich the format is based on locale

File: src/forms/medication-dispense-review.component.tsx
Comment (pirupius) [inline src/forms/medication-dispense-review.component.tsx:518]: The time (12:00AM) being displayed might be because you added `startOf('day')` here

File: src/forms/medication-dispense-review.component.test.tsx
Comment (mogoodrich) [inline src/forms/medication-dispense-review.component.test.tsx:22]: Updated, fyi @chibongho , this is much cleaner!

File: src/forms/medication-dispense-review.component.tsx
Comment (mogoodrich) [reply src/forms/medication-dispense-review.component.tsx:504]: Thanks @pirupius s 

@pirupius @mseaton  I couldn't reproduce this behavior ("Oct 12:00AM") today, I'm wondering if it had anything to do with the fact that it was the first of the month when I was testing?

I did remove the "startOf('day')" as you suggested @pirupius ... I believe that was only something I added in hopes it might fix the above issue.

But I also agree with @mseaton , we should likely start another ticket/convo about the OpenmrsDataPicker and not have it block this ticket.

**Comment by mogoodrich** [comment]
@chibongho  I fixed the mock thing, thanks!  I also made one tweak to the OpenmrsDatePickup, but do agree with @mseaton  that other issues with the datepicker should probably be ticketed as a separate ticket.  Do you think I've gotten far enough to get an approval on these changes and get them merged in?

**Review by chibongho — APPROVED** [review]

**Comment by chibongho** [comment]
> @chibongho I fixed the mock thing, thanks! I also made one tweak to the OpenmrsDatePickup, but do agree with @mseaton that other issues with the datepicker should probably be ticketed as a separate ticket. Do you think I've gotten far enough to get an approval on these changes and get them merged in?

Yeah, fine with me. Thanks!

**Review by pirupius — APPROVED** [review]
Sounds good me @mogoodrich

**Review by mogoodrich — COMMENTED** [review]
Merging, thanks all!

**Pull request merged**
